### PR TITLE
A faster evaluation for library ff by avoiding threading on serial execution

### PR DESCRIPTION
**Pull request by ceriottm**
A number of small tweaks to accelerate execution when the underlying FF is really quick, and there is a high degree of parallelism (e.g. in REMD).

Do not merge yet.

### CODE REVIEW

**Comment by ceriottm** [comment]
My, the socket code is really a mess. But with a number of targeted tweaks, it is now noticeably faster.
Would be good to test this a bit before merging.

**Comment by ceriottm** [comment]
Any objection to merging this. I tried to clean up a bit the socket interface but IMO a full rewrite is in order. Also, in retrospect, the communication protocol is really not very well thought. Very inefficient with all of that back and forth poking.

File: ipi/engine/forcefields.py [hunk 1]
```diff
@@ -330,7 +340,7 @@ def __init__(self, latency=1.0e-3, name="", pars=None, dopbc=False):
             raise ValueError("Periodic boundary conditions are not supported by FFLennardJones.")
 
         # a socket to the communication library is created or linked
-        super(FFLennardJones, self).__init__(latency, name, pars, dopbc=False)
+        super(FFLennardJones, self).__init__(latency, name, pars, dopbc=dopbc, threaded=threaded)
```
Comment (mahrossi) [inline ipi/engine/forcefields.py]: Is it intended here for dopbc=dopbc?

File: ipi/inputs/motion/motion.py [hunk 2]
```diff
@@ -179,11 +179,18 @@ def store(self, motion):
...
         if type(motion) is MultiMotion:
             self.mode.store("multi")
-            self.extra = []
-            for m in motion.mlist:
-                im = InputMotionBase()
-                im.store(m)
-                self.extra.append(("motion", im))
+
+
+            if len(self.extra) != len(motion.mlist):
```
Comment (mahrossi) [inline ipi/inputs/motion/motion.py]: @ceriottm : what is this for exactly? Seems unrelated. Is it from another branch?

**Comment by mahrossi** [comment]
Regarding the communication protocol the problem is that changing it would break most interfaces to external codes coded so far. I would refrain from doing that unless strictly necessary.

**Comment by OndrejMarsalek** [comment]
@ceriottm, I agree that the socket code is not in top condition, but did not consider it a priority, as (at least in my mind) we should be transitioning to something asyncio-based once we are on Python 3. So I agree with "a full rewrite is in order", but would rather do it using the new language features that Python 3 offers. That would also be a good opportunity to start versioning the protocol and potentially make it more flexible.

Most people are probably aware, but just for completeness, a comment on threading in Python. It is useful for the "traditional" approach to sockets (or perhaps I/O in general), but because of the GIL, you can't expect any performance gains from it for CPU-bound stuff. There are ways to work around that if we ever want to have native Python implementations of simple interactions, but threading isn't one of them.

I am looking at the code now, apologies for the hold up.

File: ipi/engine/motion/dynamics.py [hunk 3]
```diff
@@ -317,16 +317,16 @@ def pconstraints(self):
...
+                pcom /= Mnb
+                self.beads.p[:, i:na3:3] -= m * pcom
 
-            self.ensemble.eens += np.dot(pcom, pcom) / (2.0 * M * nb)
-
-            # subtracts COM velocity
-            pcom *= 1.0 / (nb * M)
-            for i in range(3):
-                self.beads.p[:, i:na3:3] -= m * pcom[i]
+            self.ensemble.eens += dens*0.5/(Mnb)
```
Comment (OndrejMarsalek) [inline ipi/engine/motion/dynamics.py]:
```suggestion
            self.ensemble.eens += dens * 0.5 / Mnb
```

File: ipi/engine/motion/dynamics.py [hunk 4]
```diff
@@ -317,16 +317,16 @@ def pconstraints(self):
             p = dstrip(self.beads.p)
             m = dstrip(self.beads.m3)[:, 0:na3:3]
             M = self.beads[0].M
+            Mnb = M*nb
 
+            dens = 0
             for i in range(3):
-                pcom[i] = p[:, i:na3:3].sum()
+                pcom = p[:, i:na3:3].sum()
```
Comment (OndrejMarsalek) [inline ipi/engine/motion/dynamics.py]: Looks like with these changes, line 313 above:
```
pcom = np.zeros(3, float)
```
is not needed.

File: ipi/engine/forces.py [hunk 5]
```diff
@@ -161,6 +161,7 @@ def get_all(self):
             self._getallcount += 1
 
         # this is converting the distribution library requests into [ u, f, v ]  lists
+        t_start = time.time()
```
Comment (OndrejMarsalek) [inline ipi/engine/forces.py]: This is never used.

File: ipi/engine/normalmodes.py [hunk 6]
```diff
@@ -265,7 +267,13 @@ def get_fspringnm(self):
     def get_vspring(self):
         """Returns the spring energy calculated in NM representation."""
 
-        vspring = (self.beads.m3 * self.omegak[:, np.newaxis]**2 * self.qnm**2).sum()
+        if self.nbeads == 0:
```
Comment (OndrejMarsalek) [inline ipi/engine/normalmodes.py]: Is there such a thing or do you mean `if self.nbeads == 1:`?

File: ipi/engine/normalmodes.py [hunk 7]
```diff
@@ -527,17 +535,11 @@ def get_kins(self):
...
-        sm = dstrip(self.beads.sm3[0])
-        pnm = dstrip(self.pnm)
-        nmf = dstrip(self.nm_factor)
-
-        # computes the MD ke in the normal modes representation, to properly account for CMD mass scaling
-        for b in range(self.nbeads):
-            sp = pnm[b] / sm                      # mass-scaled momentum of b-th NM
-            kmd[b] = np.dot(sp, sp) * 0.5 / nmf[b]   # include the partially adiabatic CMD mass scaling
+        """
+	# include the partially adiabatic CMD mass scaling
```
Comment (OndrejMarsalek) [inline ipi/engine/normalmodes.py]:
```suggestion
        # include the partially adiabatic CMD mass scaling
```

Comment (OndrejMarsalek) [reply ipi/engine/normalmodes.py]: A wild tab character appears! ;-)

File: ipi/engine/normalmodes.py [hunk 8]
```diff
@@ -527,17 +535,11 @@ def get_kins(self):
...
+        pnm = dstrip(self.pnm)/dstrip(self.beads.sm3)
```
Comment (OndrejMarsalek) [inline ipi/engine/normalmodes.py]:
```suggestion
        pnm = dstrip(self.pnm) / dstrip(self.beads.sm3)
```

File: ipi/engine/normalmodes.py [hunk 9]
```diff
@@ -527,17 +535,11 @@ def get_kins(self):
...
+        kmd = 0.5*(pnm**2).sum(axis=1)/dstrip(self.nm_factor)
```
Comment (OndrejMarsalek) [inline ipi/engine/normalmodes.py]:
```suggestion
        kmd = 0.5 * (pnm**2).sum(axis=1) / dstrip(self.nm_factor)
```

Comment (OndrejMarsalek) [reply ipi/engine/normalmodes.py]: Great to see the continuing pythonification :+1: We need more of that.

File: ipi/engine/smotion/remd.py [hunk 10]
```diff
@@ -169,3 +184,6 @@ def step(self, step=None):
                     sf.write(" % 5d" % (i))
                 sf.write("\n")
 
+        info("# REMD step evaluated in %f (%f eval, %f swap) sec." % (time.time()-t_start, t_eval, t_swap), verbosity.debug)
+
+
```
Comment (OndrejMarsalek) [inline ipi/engine/smotion/remd.py]:
```suggestion
```

File: ipi/engine/smotion/remd.py [hunk 11]
```diff
@@ -169,3 +184,6 @@ def step(self, step=None):
                     sf.write(" % 5d" % (i))
                 sf.write("\n")
 
+        info("# REMD step evaluated in %f (%f eval, %f swap) sec." % (time.time()-t_start, t_eval, t_swap), verbosity.debug)
+
```
Comment (OndrejMarsalek) [inline ipi/engine/smotion/remd.py]:
```suggestion
```

File: ipi/engine/smotion/remd.py [hunk 12]
```diff
@@ -169,3 +184,6 @@ def step(self, step=None):
                     sf.write(" % 5d" % (i))
                 sf.write("\n")
 
+        info("# REMD step evaluated in %f (%f eval, %f swap) sec." % (time.time()-t_start, t_eval, t_swap), verbosity.debug)
```
Comment (OndrejMarsalek) [inline ipi/engine/smotion/remd.py]: Ideally, we would have a more systematic timing mechanism throughout the code, I even have a timer class ready, but so far other things were more pressing. One of these days...

File: ipi/inputs/forces.py [hunk 13]
```diff
@@ -131,9 +131,17 @@ def store(self, flist):
...
+        for ii, el in enumerate(flist):
+            if self.extra[ii] == 0:
+                iff = InputForceComponent()
+                iff.store(el)
+                self.extra[ii] = ("force", iff)
+            else:
+                self.extra[ii][1].store(el)
+
+
+
```
Comment (OndrejMarsalek) [inline ipi/inputs/forces.py]:
```suggestion
```

File: ipi/inputs/forces.py [hunk 14]
```diff
@@ -131,9 +131,17 @@ def store(self, flist):
...
+        for ii, el in enumerate(flist):
+            if self.extra[ii] == 0:
+                iff = InputForceComponent()
+                iff.store(el)
+                self.extra[ii] = ("force", iff)
+            else:
+                self.extra[ii][1].store(el)
+
+
```
Comment (OndrejMarsalek) [inline ipi/inputs/forces.py]:
```suggestion
```

File: ipi/inputs/forces.py [hunk 15]
```diff
@@ -131,9 +131,17 @@ def store(self, flist):
...
+            self.extra = [0] * len(flist)
+
+        for ii, el in enumerate(flist):
+            if self.extra[ii] == 0:
+                iff = InputForceComponent()
+                iff.store(el)
+                self.extra[ii] = ("force", iff)
+            else:
+                self.extra[ii][1].store(el)
+
```
Comment (OndrejMarsalek) [inline ipi/inputs/forces.py]:
```suggestion
```

File: ipi/inputs/motion/motion.py [hunk 16]
```diff
@@ -179,11 +179,18 @@ def store(self, motion):
 
         if type(motion) is MultiMotion:
             self.mode.store("multi")
-            self.extra = []
-            for m in motion.mlist:
-                im = InputMotionBase()
-                im.store(m)
-                self.extra.append(("motion", im))
+
```
Comment (OndrejMarsalek) [inline ipi/inputs/motion/motion.py]:
```suggestion
```

File: ipi/inputs/system.py [hunk 17]
```diff
@@ -8,6 +8,7 @@
 import os.path
 import sys
 import re
+import time
```
Comment (OndrejMarsalek) [inline ipi/inputs/system.py]: Not used.

File: ipi/interfaces/sockets.py [hunk 18]
```diff
@@ -334,44 +338,45 @@ def getforce(self):
...
-        mlen = np.int32()
-        mlen = self.recvall(mlen)
-        if mlen > 0:
-            mxtra = np.zeros(mlen, np.character)
-            mxtra = self.recvall(mxtra)
-            mxtra = "".join(mxtra)
-        else:
-            mxtra = ""
+        with self.threadlock:
+            if (self.status & Status.HasData):
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
            if self.status & Status.HasData:
```

File: ipi/interfaces/sockets.py [hunk 19]
```diff
@@ -334,44 +338,45 @@ def getforce(self):
...
+                self.sendall(Message("getforce"));
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
                self.sendall(Message("getforce"))
```

File: ipi/interfaces/sockets.py [hunk 20]
```diff
@@ -552,135 +688,71 @@ def pool_distribute(self):
...
-            freec.remove(c)
+        with self.threadlock:
+            freec = self.clients[:]
+            for [r2, c] in self.jobs:
+                freec.remove(c)
 
-        # fills up list of pending requests if empty
-        if len(self.prlist) == 0:
+        # fills up list of pending requests if empty, or if clients are abundant
+        if len(self.prlist) == 0 or len(freec)>len(self.prlist):
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
        if (len(self.prlist) == 0) or (len(freec) > len(self.prlist)):
```

Comment (OndrejMarsalek) [reply ipi/interfaces/sockets.py]: Parentheses are not needed here but I find them useful for readability.

File: ipi/interfaces/sockets.py [hunk 21]
```diff
@@ -552,135 +688,71 @@ def pool_distribute(self):
...
+        ttotal = tdispatch = tcheck = tjoin = 0
+        ttotal -= time.time()
+
         # get clients that are still free
-        freec = self.clients[:]
-        for [r2, c] in self.jobs:
-            freec.remove(c)
+        with self.threadlock:
+            freec = self.clients[:]
+            for [r2, c] in self.jobs:
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
            for r2, c in self.jobs:
```

File: ipi/interfaces/sockets.py [hunk 22]
```diff
@@ -552,135 +688,71 @@ def pool_distribute(self):
...
         elif self.match_mode == "any":
             match_seq = ["any"]
 
         # first: dispatches jobs to free clients (if any!)
         # tries first to match previous replica<>driver association, then to get new clients, and only finally send the a new replica to old drivers
-        if len(freec) > 0 and len(self.prlist) > 0:
+        sthreads = []
+        ndispatch = 0
+        tdispatch -= time.time()
+        while len(freec) > 0 and len(self.prlist) > 0:
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
        while (len(freec) > 0) and (len(self.prlist) > 0):
```

File: ipi/interfaces/sockets.py [hunk 23]
```diff
@@ -552,135 +688,71 @@ def pool_distribute(self):
...
-                            freec.remove(fc)
-                            # removes r from the list of pending jobs
-                            self.prlist.remove(r)
-                            break
-                        else:
-                            warning(" @SOCKET: Client " + str(fc.peername) + " is in an unexpected status " + str(fc.status) + " at (2). Will try to keep calm and carry on.", verbosity.low)
+                    if self.dispatch_free_client(fc, match_ids, sthreads):
+                        freec.remove(fc)
+                        ndispatch += 1
+                    if len(self.prlist)==0:
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
                    if len(self.prlist) == 0:
```

File: ipi/interfaces/sockets.py [hunk 24]
```diff
@@ -552,135 +688,71 @@ def pool_distribute(self):
...
+                        break
+            if len(freec)>0:
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
            if len(freec) > 0:
```

File: ipi/interfaces/sockets.py [hunk 25]
```diff
@@ -552,135 +688,71 @@ def pool_distribute(self):
...
-        # if len(pendr)>0:
-        #   self.poll_iter = UPDATEFREQ
         # now check for client status
-        for c in self.clients:
-            if c.status == Status.Disconnected:  # client disconnected. force a pool_update
-                self.poll_iter = UPDATEFREQ
-                return
-            if not c.status & (Status.Ready | Status.NeedsInit):
-                c.poll()
+        if len(self.jobs)==0:
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
        if len(self.jobs) == 0:
```

File: ipi/interfaces/sockets.py [hunk 26]
```diff
@@ -691,7 +763,8 @@ def poll(self):
         """
 
         # makes sure to remove the last dead client as soon as possible -- and to get clients if we are dry
-        if self.poll_iter >= UPDATEFREQ or len(self.clients) == 0 or (len(self.clients) > 0 and not(self.clients[0].status & Status.Up)):
+        if (self.poll_iter >= UPDATEFREQ or len(self.clients) == 0 or
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
        if ((self.poll_iter >= UPDATEFREQ) or (len(self.clients) == 0) or
```

File: ipi/interfaces/sockets.py [hunk 27]
```diff
@@ -691,7 +763,8 @@ def poll(self):
...
+             (len(self.clients) > 0 and not(self.clients[0].status & Status.Up))):
```
Comment (OndrejMarsalek) [inline ipi/interfaces/sockets.py]:
```suggestion
             ((len(self.clients) > 0) and not(self.clients[0].status & Status.Up))):
```

File: ipi/utils/nmtransform.py [hunk 28]
```diff
@@ -129,6 +129,20 @@ def mk_o_rs_matrix(nb1, nb2):
...
         return mk_o_rs_matrix(nb2, nb1).T * (float(nb2) / float(nb1))
 
+class nm_noop(object):
+    """ A no-op NM transformation for classical trajectories """
+
+    def __init__(self, nbeads, open_paths=None):
+        """Initializes nm_noop. """
+        if nbeads > 1:
+            raise ValueError("Shouldn't use a noop transformation for ring-polymer systems ")
+
```
Comment (OndrejMarsalek) [inline ipi/utils/nmtransform.py]:
```suggestion
```

File: ipi/engine/forcefields.py [hunk 29]
```diff
@@ -316,7 +326,7 @@ class FFLennardJones(ForceField):
                          'start': starting time}.
     """
 
-    def __init__(self, latency=1.0e-3, name="", pars=None, dopbc=False):
+    def __init__(self, latency=1.0e-3, name="", pars=None, dopbc=False, threaded=False):
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/engine/forcefields.py]: Do you have changes that you'd apply here or these are more thoughts for a future clean-up?

File: ipi/engine/forcefields.py [hunk 30]
```diff
@@ -397,7 +407,7 @@ class FFQUIP(ForceField):
                          'start': starting time}.
     """
 
-    def __init__(self, init_file, args_str, param_file, latency=1.0e-3, name="", pars=None, dopbc=True):
+    def __init__(self, init_file, args_str, param_file, latency=1.0e-3, name="", pars=None, dopbc=True, threaded=False):
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/engine/forcefields.py]: So you would prevent using threaded at all? Seemed a bit brutal to me.

**Comment by mahrossi** [comment]
Ehm, I have a question: Now this is becoming a different pull request. Is this the best way to go forward? It needs to be re-reviewed, right?

And for clarification: your restructuring of the sockets code does not break compatibility with interfaces that already exist, correct?

**Comment by ceriottm** [comment]
@mahrossi  yes, it's kind of another major change, but deeply related - actually it undoes many of the changes I had made here. everything should still work - so please do test

File: ipi/engine/motion/dynamics.py (hunk 4 above)
Comment (ceriottm) [reply ipi/engine/motion/dynamics.py]: agreed.

File: ipi/engine/smotion/remd.py (hunk 12 above)
Comment (ceriottm) [reply ipi/engine/smotion/remd.py]: Agreed.

File: ipi/inputs/system.py (hunk 17 above)
Comment (ceriottm) [reply ipi/inputs/system.py]: there are many of those.... I'd let them be in case we need to time again. until we have proper timing mechanism :-)

**Comment by ceriottm** [comment]
OK this really needs to be merged. I leave the branch alive and will be glad to discuss further fixes and improvements, but I need to branch off some new developments and I'd rather do them from a merged master.
